File: patterns/javascript/DuckTypeClassDecorator.js
class Core {
  constructor() {
    this.variable = null;
  }
  get() {
    return this.variable;
  }
  set(value) {
    this.variable = value;
    return value;
  }
  toString(){
    return `(${this.value})`
  }
}

class Decorator1 {
  constructor(instance) {
    this.instance = instance;
  }
  get() {
    console.log('get method 1');
    return this.instance.get();
  }
  set(value) {
    console.log('put method 1');
    this.instance.set(value);
    return value;
  }
}

class Decorator2 {
  constructor(instance) {
    this.instance = instance;
  }
  get() {
    console.log('get method 2');
    return this.instance.get();
  }
  set(value) {
    console.log('put method 2');
    this.instance.set(value);
    return value;
  }
}

const c = new Core();
c.set('hola')
console.log(c.get());

const d1 = new Decorator1(c);
d1.set('chau');
console.log(d1.get());

const d2 = new Decorator2(c);
d2.set('ultimo');
console.log(d2.get());

File: patterns/python/DuckTypeClassDecorator.py
#! /usr/bin/env python

class Core:
    def __init__(self):
        self._variable = None

    def get(self):
        return self._variable

    def put(self, value):
        self._variable = value
        return value


class Decorator1:
    def __init__(self, instance):
        self._instance = instance

    def get(self):
        print('get method 1')
        return self._instance.get()

    def put(self, value):
        print('put method 1')
        return self._instance.put(value)


class Decorator2:
    def __init__(self, instance):
        self._instance = instance

    def get(self):
        print('get method 2')
        return self._instance.get()

    def put(self, value):
        print('put method 2')
        return self._instance.put(value)


c = Core()
c.put('hola')
print(c.get())

d1 = Decorator1(c)
d1.put('chau')
print(d1.get())

d2 = Decorator2(c)
d2.put('ultimo')
print(d2.get())
